### ISSUE TITLE
Unable to load the kernel module 'nvidia.ko' - OKD 4.7

### ISSUE THREAD

**Issue by fubsle** [NVIDIA/gpu-operator#239]
### Specs:
- OKD: 4.7.0-0.okd-2021-08-07-063045 (IPI on vSphere)
- FCOS: 34
- Kernel: 5.12.19-300.fc34.x86_64
- nvidia GPU Operator 1.7.1 installed from OperatorHub
- Node Feature Discovery 4.7.0 installed from OperatorHub
- Driver: 470.57.02

### Problem: 
nvidia-driver-daemonset is failing with the following error:
```
ERROR: Unable to load the kernel module 'nvidia.ko'.  This happens most frequently when this kernel module was built against the wrong or improperly configured kernel sources, with a version of gcc that differs from the one used to build the target kernel, or if another driver, such as nouveau, is present and prevents the NVIDIA kernel module from obtaining ownership of the NVIDIA GPU(s), or no NVIDIA GPU installed in this system is supported by this NVIDIA Linux graphics driver release.

Please see the log entries 'Kernel module load error' and 'Kernel messages' at the end of the file '/var/log/nvidia-installer.log' for more information.
```
Happens when this command is executed: 
```IGNORE_CC_MISMATCH=1 nvidia-installer --kernel-module-only --no-drm --ui=none --no-nouveau-check ${install_args[@]+"${install_args[@]}"}```

Some lines above that I found this - not sure if relevant to this issue:
```
Skipping BTF generation for /usr/src/nvidia-470.57.02/kernel/nvidia-peermem.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-470.57.02/kernel/nvidia-drm.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-470.57.02/kernel/nvidia-modeset.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-470.57.02/kernel/nvidia-uvm.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-470.57.02/kernel/nvidia.ko due to unavailability of vmlinux
```
Happens when this command is executed: 
```make -s -j SYSSRC=/lib/modules/${KERNEL_VERSION}/build nv-linux.o nv-modeset-linux.o > /dev/null```

### Setup
We are using as basis the files from https://gitlab.com/nvidia/container-images/driver/-/tree/master/fedora. 
[nvidia-driver](https://github.com/NVIDIA/gpu-operator/files/6969388/nvidia-driver.txt) has been customized in order to 
- fix gcc issue, as described in https://gitlab.com/nvidia/container-images/driver/-/issues/16
- get the correct kernel-devel, kernel-headers, kernel-core
- as well as workaround the mentioned issue in https://github.com/NVIDIA/gpu-operator/issues/231

### Further debug information
- Complete Log file: [complete.log](https://github.com/NVIDIA/gpu-operator/files/6970072/complete.log)
- A nvidia GPU is attached to the node, otherwise daemonset would not schedule the pod.
```
sh-5.1# cat /proc/version
Linux version 5.12.19-300.fc34.x86_64 (mockbuild@bkernel01.iad2.fedoraproject.org) (gcc (GCC) 11.1.1 20210531 (Red Hat 11.1.1-3), GNU ld version 2.35.1-41.fc34) #1 SMP Tue Jul 20 22:07:38 UTC 2021
sh-5.1# uname -a
Linux nvidia-driver-daemonset-2nvwl-debug 5.12.19-300.fc34.x86_64 #1 SMP Tue Jul 20 22:07:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

### Steps to reproduce the issue
1. clone https://gitlab.com/nvidia/container-images/driver/-/tree/master/fedora
2. replace nvidia-driver with attached one
3. Use ```--build-arg FEDORA_VERSION=34 --build-arg DRIVER_VERSION=470.57.02``` when building Image

**Comment by shivamerla** [comment]
@fubsle Only `470.57.02` fails and `460.73.01` works?

**Comment by fubsle** [comment]
By using the Dockerfile and nvidia-driver ```--build-arg DRIVER_VERSION=460.73.01``` another error comes up:
```
+ trap 'make -s -j SYSSRC=/lib/modules/5.12.19-300.fc34.x86_64/build clean > /dev/null' EXIT
+ echo 'Compiling NVIDIA driver kernel modules...'
+ cd /usr/src/nvidia-460.73.01/kernel
+ make -s -j SYSSRC=/lib/modules/5.12.19-300.fc34.x86_64/build nv-linux.o nv-modeset-linux.o
/usr/src/nvidia-460.73.01/kernel/nvidia-drm/nvidia-drm-modeset.c: In function '__will_generate_flip_event':
/usr/src/nvidia-460.73.01/kernel/nvidia-drm/nvidia-drm-modeset.c:96:23: warning: unused variable 'primary_plane' [-Wunused-variable]
   96 |     struct drm_plane *primary_plane = crtc->primary;
      |                       ^~~~~~~~~~~~~
Skipping BTF generation for /usr/src/nvidia-460.73.01/kernel/nvidia-modeset.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-460.73.01/kernel/nvidia-drm.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-460.73.01/kernel/nvidia-uvm.ko due to unavailability of vmlinux
Skipping BTF generation for /usr/src/nvidia-460.73.01/kernel/nvidia.ko due to unavailability of vmlinux
ld: cannot find nvidia/nv-interface.o: No such file or directory
ld: cannot find nvidia-modeset/nv-modeset-interface.o: No such file or directory
make: *** [Makefile:113: nv-linux.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [Makefile:113: nv-modeset-linux.o] Error 1
++ make -s -j SYSSRC=/lib/modules/5.12.19-300.fc34.x86_64/build clean
Stopping NVIDIA persistence daemon...
Unloading NVIDIA driver kernel modules...
Unmounting NVIDIA driver rootfs...
```

**Comment by takyon77** [comment]
@fubsle Hi, I'm having the same issue, I'm getting a 'permission denied' when trying to insert the nvidia kernel module.
I was able to insert this module on the server running the container (but outside the container of course) so the module itself seems ok. I also tried to compile a simple 'hello world' kernel module in the container and got the same 'permission denied' when trying to insert the module. So apparently on OKD 4.7, it's impossible to insert any kernel module for a reason I'm still not able to determine. 
Did you find any solution?

**Comment by takyon77** [comment]
So for anyone interested, I managed to solve my issue. It was in fact caused by a selinux 'bug' on the host running the container. Apparently a new lockdown class was introduced for selinux in Fedora 34 and this was an unwanted side-effect (https://www.spinics.net/lists/fedora-devel/msg284833.html). I just created a selinux policy on the host running the container to go around the problem and now I can insert kernel modules from the container.
